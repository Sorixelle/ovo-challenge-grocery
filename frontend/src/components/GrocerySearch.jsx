import React, { useState, useEffect } from 'react'
import http from '../http'

const GrocerySearch = ({ items, updateItems, hiddenItems, setHidden }) => {
  const [input, setInput] = useState('')
  const [categoryFilter, setCategoryFilter] = useState('')

  const [newQuantity, setNewQuantity] = useState(1)
  const [newCategory, setNewCategory] = useState('')

  const categories = [...new Set(items.map(i => i.category).filter(i => i !== null))]

  useEffect(() => {
    setHidden(items
      .map(i => ({
        ...i,
        category: i.category === null ? '' : i.category
      }))
      .filter(i =>
        !i.item.toLowerCase().startsWith(input.toLowerCase()) ||
        (categoryFilter !== '' && i.category !== categoryFilter)
      )
      .map(i => i.item)
    )
  }, [input, categoryFilter])

  function handleDropdown (e) {
    setCategoryFilter(
      e.target.value === 'Filter categories...' ? '' : e.target.value
    )
  }

  async function submitItem () {
    const item = {
      item: input,
      quantity: parseInt(newQuantity)
    }
    if (newCategory !== '') {
      item.category = newCategory
    }

    const { data } = await http.post('/add', item)

    updateItems([...items, data])

    setInput('')
    setNewQuantity(1)
    setNewCategory('')
  }

  const itemNames = new Set(items.map(i => i.item))
  const addMode = hiddenItems.length === itemNames.size && hiddenItems.every(i => itemNames.has(i))

  return (
    <form className='input-group input-group-lg' onSubmit={submitItem}>
      {addMode
        ? (
          <div style={{ display: 'contents' }}>
            <input
              type='number' style={{ width: '15%' }}
              value={newQuantity} onChange={e => setNewQuantity(e.target.value)}
            />
            <input
              type='text' placeholder='Category'
              value={newCategory} onChange={e => setNewCategory(e.target.value)}
            />
          </div>
        ) : (
          <div className='input-group-prepend'>
            <label className='input-group-text' htmlFor='searchAddField'>
              Search
            </label>
          </div>
        )}
      <input
        id='searchAddField' className='form-control'
        type='text' value={input}
        onChange={e => setInput(e.target.value)}
      />
      <div className='input-group-append'>
        {addMode
          ? (
            <button
              className='btn btn-success' type='button'
              onClick={submitItem}
            >
              Add
            </button>
          ) : (
            <select onChange={handleDropdown} value={categoryFilter}>
              <option>Filter categories...</option>
              {categories.map(c => (
                <option key={c}>{c}</option>
              ))}
            </select>
          )}
      </div>
    </form>
  )
}

export default GrocerySearch
