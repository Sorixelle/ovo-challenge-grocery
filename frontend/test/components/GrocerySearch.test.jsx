import React from 'react'
import { render } from '@testing-library/react'
import GrocerySearch from '../../src/components/GrocerySearch'

describe('GrocerySearch', () => {
  const items = [
    {
      item: 'Test',
      category: 'Test Items',
      quantity: 1
    },
    {
      item: 'Test 2',
      category: 'Other Items',
      quantity: 1
    },
    {
      item: 'Test 3',
      category: null,
      quantity: 1
    }
  ]
  const node = <GrocerySearch items={items} hiddenItems={[]} setHidden={() => {}} />

  it('should match the snapshot', () => {
    const { container } = render(node)

    expect(container).toMatchSnapshot()
  })
})
