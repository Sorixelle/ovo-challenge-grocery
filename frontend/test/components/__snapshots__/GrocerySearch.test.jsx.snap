// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GrocerySearch should match the snapshot 1`] = `
<div>
  <form
    class="input-group input-group-lg"
  >
    <div
      class="input-group-prepend"
    >
      <label
        class="input-group-text"
        for="searchAddField"
      >
        Search
      </label>
    </div>
    <input
      class="form-control"
      id="searchAddField"
      type="text"
      value=""
    />
    <div
      class="input-group-append"
    >
      <select>
        <option>
          Filter categories...
        </option>
        <option>
          Test Items
        </option>
        <option>
          Other Items
        </option>
      </select>
    </div>
  </form>
</div>
`;
